src/Topshelf/Runtime/Windows/ServiceRecoveryOptions.cs

```diff
@@ -34,6 +34,12 @@ public IEnumerable<ServiceRecoveryAction> Actions
 
         public void AddAction(ServiceRecoveryAction serviceRecoveryAction)
         {
+            if (_actions.Count == 3)
+            {
+                throw new TopshelfException("Recovery action can not be added. " +
+                                            "Windows recovery does not support more than 3 actions");
+            }
+

```

Comment:
Here we can explicitly limit number of actions.

Reply:
FYI: Windows supports more than 3 actions when configured directly with sc.exe:
`sc failure SERVICE actions=restart/1000/restart/2000/restart/5000/restart/10000/restart/30000/restart/60000/restart/300000/reboot/300000/reboot/1800000/reboot/7200000`
Tested in Windows Server 2016 and Windows 11.